src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/Marshal/PInvokeErrorMessageTests.cs

```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using Xunit;
+
+namespace System.Runtime.InteropServices.Tests
+{
+    public class PInvokeErrorMessageTests
+    {
+        public static IEnumerable<object[]> GetErrorCode_TestData()
+        {
+            yield return new object[] { 0 };
+            yield return new object[] { 1 };
+
+            // errno values
+            yield return new object[] { 0x10002 };
+            yield return new object[] { 0x10003 };
+            yield return new object[] { 0x10014 };
+            yield return new object[] { 0x1001D };
+
+            // HRESULT values
+            yield return new object[] { unchecked((int)0x80004001) };
+            yield return new object[] { unchecked((int)0x80004005) };
+            yield return new object[] { unchecked((int)0x80070057) };
+            yield return new object[] { unchecked((int)0x8000FFFF) };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetErrorCode_TestData))]
+        public void PInvokeErrorMessage_Returns_Win32Exception_Message(int error)
+        {
+            // The Win32Exception represents the canonical system exception on
+            // all platforms. The GetPInvokeErrorMessage API is about providing
+            // this message in a manner that avoid the instantiation of a Win32Exception
+            // instance and querying for the message.
+            string expected = new Win32Exception(error).Message;
```

Comment:
This test probably isn't enough as it's just testing one method calls the other. The messages should be super stable, so it should be possible to just verify that the message is the expected even if we only test the message for one error code on each platform.

Reply:
> This test probably isn't enough as it's just testing one method calls the other. 

That isn't what this test is testing. It is verifying our long-standing pattern continues to work and people can use this new API and get the same semantics.

> The messages should be super stable, so it should be possible to just verify that the message is the expected even if we only test the message for one error code on each platform.

I'm not sure how that is possible without hardcoding OS level error messages, which can change. Open to prior art that shows how this is done in libraries.

Reply:
I don't actually see any existing test of Win32Exception verifying that it uses FormatMessage or equivalent. So such a test can be here.

Reply:
The ideal test would call FormatMessage/strerror and compare the result. As you point out, this hole predates your change, so you need not fix it today. It would just become a hole in Marshal tests rather than Win32Exception tests..

Reply:
A minimal test would probably verify that the string was not empty, and perhaps that the strings for two codes were not the same.